### ISSUE TITLE
Inference for Literal arguments not handled 

### ISSUE THREAD

**Issue by Dr-Irv** [python/mypy#11003]
**Bug Report**

If you have a function with `Literal` arguments, and a variable that is set to one of the options, an error is reported.

**To Reproduce**
```python
from typing import Literal


def myfun(a: Literal["a", "b", "c"]) -> str:
    return a


s = "a"

myfun(s)


def myfun2(a: Literal[1, 2]) -> int:
    return a


b = 2
myfun2(b)

c = 1 if s == "a" else 2
myfun2(c)
```
reports
```
literalinf.py:10: error: Argument 1 to "myfun" has incompatible type "str"; expected "Union[Literal['a'], Literal['b'], Literal['c']]"
literalinf.py:18: error: Argument 1 to "myfun2" has incompatible type "int"; expected "Union[Literal[1], Literal[2]]"
literalinf.py:21: error: Argument 1 to "myfun2" has incompatible type "int"; expected "Union[Literal[1], Literal[2]]"
```

**Expected Behavior**

No error.   The values can be inferred to be of the `Literal` value

**Actual Behavior**

See above.

**Your Environment**

- Mypy version used:  0.910
- Mypy command-line flags:   None
- Mypy configuration options from `mypy.ini` (and other config files):  None
- Python version used:   3.8.5
- Operating system and version:  Windows 10 21H1

**Comment by ethanhs** [comment]
The problem is if the variable is re-assigned to, then it could be incorrect to pass it to these functions. The solution for your scenario is to type these assignments with Final, like `a: Final = "a"`.

**Comment by erictraut** [comment]
It shouldn't be necessary to mark them as Final. If a variable is reassigned, the type checker knows it's reassigned and can flag it as an error.

There is no type violation, so these errors are false positives.

The above code type checks without errors in pyright. I don't see why it shouldn't work with mypy.

**Comment by Dr-Irv** [comment]
> It shouldn't be necessary to mark them as Final. If a variable is reassigned, the type checker knows it's reassigned and can flag it as an error.
> 
> There is no type violation, so these errors are false positives.
> 
> The above code type checks without errors in pyright. I don't see why it shouldn't work with mypy.

I agree.  @ethanhs based on @erictraut 's comment, can you reopen this?

**Comment by Zac-HD** [comment]
I get a related error when setting the default argument to a function:

```python
from typing import Literal, Sequence

def f(arg: Sequence[Literal["a"]] = ("a",)) -> None:
    pass
```
```
example.py:3:37: error: Incompatible default for argument "arg"
    (default has type "Tuple[str]", argument has type "Sequence[Literal['a']]")
```

**Comment by hauntsaninja** [comment]
Hello @Zac-HD ! I enjoyed your recent paper :-) The default arguments issue is fixed on master, in #11236

**Comment by Zac-HD** [comment]
Thanks, and thanks!  I'm really looking forward to 0.920 and the release after 😁
